ddtrace/propagation/http.py

```diff
@@ -483,10 +492,214 @@ def _extract(headers):
+        not present or less than or equal to 0, update the sampling priority field to 1.
+
+        3. If the traceparent sampled flag == 1, and the tracestate sampling priority is greater than 0,
+        use the extracted sampling priority value.
+
+        4. If the traceparent sampled flag == 0, and the tracestate sampling priority is less than or equal to 0,
+        use the extracted sampling priority value.
+        """
+
+        if sampling_priority_tp == 0 and (not sampling_priority_ts or sampling_priority_ts >= 0):
+            sampling_priority = 0
+
+        elif sampling_priority_tp == 1 and (not sampling_priority_ts or sampling_priority_ts < 0):
+            sampling_priority = 1
+
+        elif sampling_priority_tp == 1 and sampling_priority_ts > 0:
+            sampling_priority = sampling_priority_ts
+
+        elif sampling_priority_tp == 0 and sampling_priority_ts < 0:
+            sampling_priority = sampling_priority_ts
+
+        return sampling_priority
+
+    @staticmethod
+    def _extract(headers):
+        # type: (Dict[str, str]) -> Optional[Context]
+
+        try:
+            tp = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACEPARENT, headers)
+            if tp is None:
+                return None
+            traceparent_values = _TraceContext._get_traceparent_values(tp)
+        except (ValueError, AssertionError):
+            log.exception("received invalid w3c traceparent: %s ", tp)
+            return None
+
+        trace_id, span_id, sampling_priority = traceparent_values
+        meta = {"traceparent": tp}
+
+        ts = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACESTATE, headers)
+        if ts:
+            # the value MUST contain only ASCII characters in the
+            # range of 0x20 to 0x7E except comma (,) and equal-sign (=)
+            if re.search(r"[^\x20-\x7E]+", ts):
+                log.debug("received invalid tracestate header: %r", ts)
+            else:
+                # store ts so we keep other vendor data, even if dd ends up being invalid
+                meta[_TRACESTATE_KEY] = ts
+                try:
+                    tracestate_values = _TraceContext._get_tracestate_values(ts)
+                except (TypeError, ValueError, AttributeError):
+                    log.debug(("received invalid dd header value in tracestate: %r "), ts)
+                    tracestate_values = None
+
+                if tracestate_values:
+                    sampling_priority_ts, other_propagated_tags, origin = tracestate_values
+                    meta.update(other_propagated_tags)
+
+                    sampling_priority = _TraceContext._get_sampling_priority(sampling_priority, sampling_priority_ts)
+
+                    return Context(
+                        trace_id=trace_id,
+                        span_id=span_id,
+                        sampling_priority=sampling_priority,
+                        dd_origin=origin,
+                        meta=cast(_MetaDictType, meta),
```

Comment:
I don't think we should have to do a cast here. Can we declare the type of meta above when it's initialized?

Reply:
Ah ok, I was following what other propagators did, let me give that a try.

Reply:
Oh nvm, one does that, the other sets the type when it inits like you're recommending. Changed it!